src/core/raster/qgsrasterdataprovider.cpp

```diff
@@ -647,3 +651,146 @@ QString QgsRasterDataProvider::colorInterpretationName( int bandNo ) const
 {
   return colorName( colorInterpretation( bandNo ) );
 }
+
+QgsRasterDataProvider::VirtualRasterParameters QgsRasterDataProvider::decodeVirtualRasterProviderUri( const QString &uri, bool *ok )
+{
+  QUrl url = QUrl::fromEncoded( uri.toLatin1() );
+  const QUrlQuery query( url.query() );
+  VirtualRasterParameters components;
+
+  if ( ! query.hasQueryItem( QStringLiteral( "crs" ) ) )
+  {
+    QgsDebugMsg( "crs is missing" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  if ( ! components.crs.createFromString( query.queryItemValue( QStringLiteral( "crs" ) ) ) )
+  {
+    QgsDebugMsg( "failed to create crs" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+
+
+  if ( ! query.hasQueryItem( QStringLiteral( "extent" ) ) )
+  {
+    QgsDebugMsg( "extent is missing" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  QStringList pointValuesList = query.queryItemValue( QStringLiteral( "extent" ) ).split( ',' );
+  if ( pointValuesList.size() != 4 )
+  {
+    QgsDebugMsg( "the extent is not correct" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  components.extent = QgsRectangle( pointValuesList.at( 0 ).toDouble(), pointValuesList.at( 1 ).toDouble(),
+                                    pointValuesList.at( 2 ).toDouble(), pointValuesList.at( 3 ).toDouble() );
+
+  if ( ! query.hasQueryItem( QStringLiteral( "width" ) ) )
+  {
+    QgsDebugMsg( "width is missing" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  bool flagW;
+  components.width = query.queryItemValue( QStringLiteral( "width" ) ).toInt( & flagW );
+  if ( !flagW ||  components.width < 0 )
+  {
+    QgsDebugMsg( "invalid or negative width input" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+
+  if ( ! query.hasQueryItem( QStringLiteral( "height" ) ) )
+  {
+    QgsDebugMsg( "height is missing" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  bool flagH;
+  components.height = query.queryItemValue( QStringLiteral( "height" ) ).toInt( & flagH );
+  if ( !flagH ||  components.height < 0 )
+  {
+    QgsDebugMsg( "invalid or negative width input" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+
+  if ( ! query.hasQueryItem( QStringLiteral( "formula" ) ) )
+  {
+    QgsDebugMsg( "formula is missing" );
+    if ( ok ) *ok = false;
+    return components;
+  }
+  components.formula = query.queryItemValue( QStringLiteral( "formula" ) );
+
+  for ( const auto &item : query.queryItems() )
+  {
+    if ( item.first.indexOf( ':' ) == -1 )
+    {
+      continue;
```

Comment:
can you document here when this happens?

Reply:
Yes, it checks if the uri has or not a first element of the derived `QUrlQuery query` that contains `:` such as `dem:uri` (but in this case also `dem:provider`).
In fact the code at the begininng was different in this part so I think that the second check (the one you asked in the next comment) is enough.